metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/IncomingRecordToEdmTransformer.java

```diff
@@ -0,0 +1,148 @@
+    documentBuilderFactory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+    documentBuilderFactory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+    documentBuilderFactory.setNamespaceAware(true);
+  }
+
+  /**
+   * Transform the given xmlRecord.
+   *
+   * @param xmlRecord The incoming xmlRecord (that comes from the vocabulary).
+   * @param resourceId The xmlRecord ID of the incoming xmlRecord.
+   * @return The EDM xmlRecord, or null if the xmlRecord couldn't be transformed.
+   * @throws BadContentException if there was a problem performing the transformation.
+   */
+  public Optional<String> transform(String xmlRecord, String resourceId) throws BadContentException {
+    // Set up the transformer
+    final Source source = new StreamSource(new StringReader(xmlRecord));
+    final StringWriter transformedXmlWriter = new StringWriter();
+    final Transformer transformer;
+    try {
+      transformer = template.newTransformer();
+      transformer.setParameter(TARGET_ID_PARAMETER_NAME, resourceId);
+      transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+      transformer.setOutputProperty(OutputKeys.ENCODING, StandardCharsets.UTF_8.name());
+
+      // Perform the transformation.
+      transformer.transform(source, new StreamResult(transformedXmlWriter));
+    } catch (TransformerException e) {
+      throw new BadContentException("Transformation failure", e);
+    }
+    return getValidatedXml(resourceId, transformedXmlWriter.toString());
+  }
+
+  /**
+   * Returns an empty optional which is empty if the provided xml is a validated empty xml or contains the xml itself if it's a
+   * valid parsable xml.
+   *
+   * @param resourceId the resource id
+   * @param xml the xml
+   * @return the optional being empty or with the xml contents
+   * @throws BadContentException if the xml parsing failed
+   */
+  @NotNull
+  private Optional<String> getValidatedXml(String resourceId, String xml) throws BadContentException {
+    final Optional<String> xmlResponse;
+    if (isEmptyXml(xml)) {
+      xmlResponse = Optional.empty();
+      if (LOGGER.isInfoEnabled()) {
+        LOGGER.info("Transformed entity {} results to an empty XML.",
+            CRLF_PATTERN.matcher(resourceId).replaceAll(""));
+      }
+    } else {
+      try {
+        isValidXml(xml);
+        xmlResponse = Optional.of(xml);
+      } catch (ParserConfigurationException | IOException | SAXException e) {
+        throw new BadContentException("Transformed xml is not valid", e);
+      }
+    }
+
+    return xmlResponse;
+  }
+
+  /**
+   * Validates if the provided xml can be parsed.
+   *
+   * @param xml the xml string
+   * @throws ParserConfigurationException if xml parsing failed
+   * @throws IOException if xml parsing failed
+   * @throws SAXException if xml parsing failed
+   */
+  private void isValidXml(String xml) throws ParserConfigurationException, IOException, SAXException {
```

Comment:
The naming for this type of method is a bit weird. At first I thought it was a `boolean` method type, but then I had to look at it again to see that it was actually `void`. Also the javadoc gives the impression that this is to return a `boolean` type. 
Looking at how it is right now, how will the user know it is valid or not? Perhaps we could rename this to something else?